blockstore/src/main/java/brooklyn/location/blockstore/AbstractVolumeManager.java

```diff
@@ -98,12 +97,36 @@ public void createFilesystem(AttachedBlockDevice attachedDevice, FilesystemOptio
             String driveLetterParam = Strings.isNullOrEmpty(driveLetter) ? "-AssignDriveLetter" : "-DriveLetter " + driveLetter;
             String volumeLabel = filesystemOptions.getVolumeLabel();
             String volumeLabelParam = Strings.isNullOrEmpty(volumeLabel) ? "datadisk" : volumeLabel;
-            WinRmToolResponse response = ((WinRmMachineLocation)machine).executePsScript("Get-Disk | Where partitionstyle -eq 'raw' | " +
+            StringBuilder builder = new StringBuilder();
+            String command =
+                    "do {" +
+                    "  $job = Get-Disk -AsJob; " +
+                    "  Wait-Job $job; " +
+                    "  $disks = Receive-Job -Job $job -Wait; " +
                     // AWS May create an instance store volume, which we need to exclude
                     // See https://docs.aws.amazon.com/AWSEC2/latest/WindowsGuide/ec2-windows-volumes.html#instance-store-volume-map
-                    "Where {$_.SerialNumber -As [int] -NotIn 78..89 -Or $_.FriendlyName -Ne \"AWS PVDISK\"} | " +
-                    "Initialize-Disk -PartitionStyle MBR -PassThru | New-Partition " + driveLetterParam + " -UseMaximumSize | " +
-                    "Format-Volume -FileSystem " + filesystemOptions.getFilesystemType() +" -NewFileSystemLabel \"" + volumeLabelParam + "\" -Confirm:$false");
+                    "  $rawDisk = $disks | Where PartitionStyle -eq 'RAW' | Where {$_.SerialNumber -As [int] -NotIn 78..89 -Or $_.FriendlyName -Ne \"AWS PVDISK\"} | Select-Object -first 1; " +
+                    "  $count = ($rawDisk | measure).Count; " +
+                    "} while ($count -lt 1); " + // Wait until the device has been attached
+                    "$job = Initialize-Disk -InputObject $rawDisk -PartitionStyle MBR -PassThru -AsJob; " +
+                    "Wait-Job $job; " +
+                    "$disk = Receive-Job -Job $job -Wait; " +
+                    "$job = New-Partition -InputObject $disk " + driveLetterParam + " -UseMaximumSize -AsJob; " +
+                    "Wait-Job $job; " +
+                    "$partition = Receive-Job -Job $job -Wait; " +
+                    "$job = Format-Volume -DriveLetter $partition.DriveLetter -FileSystem " + filesystemOptions.getFilesystemType() +" -NewFileSystemLabel \"" + volumeLabelParam + "\" -Confirm:$false -AsJob; " +
+                    "Wait-Job $job; " +
+                    "Receive-Job -Job $job -Wait";
+            builder.append("Initializing disk:\n");
+            builder.append(command);
+            WinRmToolResponse response = ((WinRmMachineLocation)machine).executePsScript(command);
```

Comment:
(minor) I note that you are building a log string with the execution in the middle, and then issuing the  LOG just once at the end. Recommend you do 2 LOG statements, one before the execution and one after. That way if `executePsScript` exceptions, or something else unusual happens, you have at least half of the log message in the log!

Reply:
Oops, that was my intention!